lib/HLSL/HLOperationLower.cpp

```diff
@@ -4173,79 +4150,113 @@ Value *TranslateBufLoad(ResLoadHelper &helper, HLResource::Kind RK,
   OP::OpCode opcode = helper.opcode;
   Type *Ty = helper.retVal->getType();
 
+  unsigned NumComponents = 1;
+  if (Ty->isVectorTy())
+    NumComponents = Ty->getVectorNumElements();
+
   const bool isTyped = DXIL::IsTyped(RK);
   Type *EltTy = Ty->getScalarType();
   const bool is64 = (EltTy->isIntegerTy(64) || EltTy->isDoubleTy());
   const bool isBool = EltTy->isIntegerTy(1);
-  Type *MemTy = EltTy;
+  // Values will be loaded in memory representations.
   if (isBool || (is64 && isTyped))
-    // Value will be loaded in its memory representation.
-    MemTy = Builder.getInt32Ty();
-  unsigned LdSize = DL.getTypeAllocSize(MemTy);
+    EltTy = Builder.getInt32Ty();
 
-  unsigned numComponents = 1;
-  Type *MaybeVecMemTy = MemTy;
-  if (Ty->isVectorTy()) {
-    numComponents = Ty->getVectorNumElements();
-    MaybeVecMemTy = VectorType::get(EltTy, numComponents);
-  }
+  unsigned LdSize = DL.getTypeAllocSize(EltTy);
 
-  std::vector<Value *> elts(numComponents);
+  std::vector<Value *> Elts(MemNumComponents);
```

Comment:
This could be `SmallVector<Value *, 4>` and it would rarely need to allocate any heap (only for long vectors).

Reply:
Wait, make it 16 and it covers matrix too.

Reply:
Scratch previous. Done.